src/main/java/com/wanderersoftherift/wotr/core/OnHitEvent.java

```diff
@@ -10,9 +11,28 @@
 public class OnHitEvent {
 
     @SubscribeEvent
-    public static void onHit(LivingIncomingDamageEvent event) {
-        float dmg = OnHitEffect.critical(event.getAmount(), (LivingEntity) event.getSource().getDirectEntity(),
-                event.getEntity(), event.getEntity().getRandom());
-        event.setAmount(dmg);
+    public static void onCriticalHit(LivingIncomingDamageEvent event) {
+        if (event.getSource().getDirectEntity() instanceof LivingEntity livingAttacker) {
+            float dmg = OnHitEffect.critical(event.getAmount(), livingAttacker, event.getEntity(),
+                    event.getEntity().getRandom());
+            event.setAmount(dmg);
+        }
+    }
+
+    @SubscribeEvent
+    public static void onThornsProc(LivingIncomingDamageEvent event) {
+        if (event.getEntity() instanceof LivingEntity living) {
+            if (living.getLastAttacker() instanceof LivingEntity attacker) {
+                boolean t = OnHitEffect.thorns(living, living.getRandom());
+                double thornsDamage = living.getAttributeValue(WotrAttributes.THORNS_DAMAGE);
```

Comment:
Just noting there probably needs to be a check somewhere for whether the the causing damage is thorns, so that thorns doesn't trigger off thorns (resulting in either the game crashing or both entities exploding).